src/SIL.LCModel/DomainServices/WritingSystemServices.cs

```diff
@@ -1524,14 +1525,54 @@ internal static void UpdateWritingSystemListField(LcmCache cache, ICmObject obj,
 			cache.DomainDataByFlid.set_UnicodeProp(obj.Hvo, flid, newVal.Length == 0 ? null : newVal);
 		}
 
+		/// <param name="cache"/>
+		/// <param name="wholeStringWSs">the handles of writing systems that have whole strings (that could have slices)</param>
+		/// <param name="embeddedWSs">the handles of writing systems that have data embedded in other writing systems' strings</param>
+		/// <returns>the handles of all Writing Systems that have data in the project</returns>
+		public static ISet<int> FindAllWritingSystemsWithData(LcmCache cache, out ISet<int> wholeStringWSs, out ISet<int> embeddedWSs)
+		{
+			var strHandles = new HashSet<int>();
+			var multiStrHandles = new HashSet<int>();
+			var inMultiStrHandles = new HashSet<int>();
+			StringServices.CrawlStrings(cache, str => FindAllWritingSystemsInTsString(str, strHandles), multiStr =>
```

Comment:
non-multi-strings could be part of:
* cell (column filter)—who cares?
* StTxtPara—has PreferredWsId ("en")
* PunctuationForm (has PreferredWsID; do we care?)

Reply:
Perhaps we should return only a single list, especially for MVP.

Reply:
For monolingual strings, determining the "outer" WS is more trouble than it's presently worth (and may be impossible). The other helpful distinction would be vernacular vs analysis, but that'd also be difficult to determine in our countless cases. We need this to offer a feature as a convenience to users who may have hidden WS's and now want to re-show them. Users ought to be able to guess whether a WS should be vern or analysis (and if data is in the wrong place, that's a separate problem).